internal/olm/operator/internal/configmap.go

```diff
@@ -0,0 +1,151 @@
+	pkgName := pkg.PackageName
+	nn := types.NamespacedName{
+		Name:      getRegistryConfigMapName(pkgName),
+		Namespace: namespace,
+	}
+	configmap := corev1.ConfigMap{}
+	err := m.Client.KubeClient.Get(ctx, nn, &configmap)
+	if err != nil {
+		return false, err
+	}
+	// Collect digests of manifests submitted to m.
+	newData, err := createConfigMapBinaryData(pkg, m.Manifests.GetBundles())
+	if err != nil {
+		return false, errors.Wrap(err, "error creating binary data")
+	}
+	// If lengths don't match we have added or removed a file.
+	if len(newData) != len(configmap.BinaryData) {
+		return true, nil
+	}
+	// Check each binary value's key, which contains a base32-encoded md5 digest
+	// component, against the new set of manifest keys.
+	for fileKey := range configmap.BinaryData {
+		if _, match := newData[fileKey]; !match {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// hashContents creates a base32-encoded md5 digest of b's bytes.
+func hashContents(b []byte) string {
+	h := md5.New()
+	h.Write(b)
+	enc := base32.StdEncoding.WithPadding(base32.NoPadding)
+	return enc.EncodeToString(h.Sum(nil))
+}
+
+// getObjectFileName opaquely creates a unique file name based on data in b.
+func getObjectFileName(b []byte, name, kind string) string {
+	digest := hashContents(b)
+	return fmt.Sprintf("%s.%s.%s.yaml", digest, name, strings.ToLower(kind))
+}
+
+func getPackageFileName(b []byte, name string) string {
+	digest := hashContents(b)
+	return fmt.Sprintf("%s.%s.package.yaml", digest, name)
+}
+
+// createConfigMapBinaryData opaquely creates a set of paths using data in pkg
+// and each bundle in bundles, unique by path. These paths are intended to
+// be keys in a ConfigMap.
+func createConfigMapBinaryData(pkg registry.PackageManifest, bundles []*registry.Bundle) (map[string][]byte, error) {
+	pkgName := pkg.PackageName
+	binaryKeyValues := map[string][]byte{}
+	pb, err := yaml.Marshal(pkg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error marshalling package manifest %s", pkgName)
+	}
+	binaryKeyValues[getPackageFileName(pb, pkgName)] = pb
+	for _, bundle := range bundles {
+		for _, o := range bundle.Objects {
+			ob, err := yaml.Marshal(o)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error marshalling object %s %q", o.GroupVersionKind(), o.GetName())
+			}
+			binaryKeyValues[getObjectFileName(ob, o.GetName(), o.GetKind())] = ob
+		}
+	}
+	return binaryKeyValues, nil
+}
+
+const (
+	// The directory containing all manifests for an operator, with the
+	// package manifest being top-level.
+	containerManifestsDir = "/registry/manifests"
+)
+
+func getRegistryConfigMapName(pkgName string) string {
+	name := k8sutil.FormatOperatorNameDNS1123(pkgName)
+	return fmt.Sprintf("%s-registry-bundles", name)
+}
+
+// withBinaryData returns a function that creates entries in the ConfigMap
+// argument's binaryData for each key and []byte value in kvs.
+func withBinaryData(kvs map[string][]byte) func(*corev1.ConfigMap) {
```

Comment:
👍 I really like that you tried to reduce the complexity by doing the encapsulations.

Reply:
Thanks!